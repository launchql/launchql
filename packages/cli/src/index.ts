#!/usr/bin/env node
import { CLI, CLIOptions } from 'inquirerer';

import { commands } from './commands';

export const options: Partial<CLIOptions> = {
  minimistOpts: {
    alias: {
      v: 'version',
      h: 'help'
    }
  }
};

const app = new CLI(commands, options);

app.run().then(()=> {
  // all done!
}).catch(error => {
  // Should not reach here with the new CLI error handling pattern
  // But keep as fallback for unexpected errors
  console.error('Unexpected error:', error);
  process.exit(1);
});