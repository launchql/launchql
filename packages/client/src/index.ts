import { Pool, PoolClient } from 'pg';

import env from './env';

export class Database {
  private static instance: Database;
  private pool: Pool;

  constructor() {
    if (Database.instance) {
      return Database.instance;
    }

    const pgPoolConfig = {
      connectionString: env.DATABASE_URL,
    };
    this.pool = new Pool(pgPoolConfig);

    // Ensure the pool is closed on process termination
    process.on('SIGTERM', async () => {
      await this.shutdown();
    });

    Database.instance = this;
    return this;
  }

  /**
   * Executes a callback function within a database transaction.
   * @param fn - A callback function that receives a PoolClient to perform database operations.
   */
  async withTransaction(fn: (client: PoolClient) => Promise<void>): Promise<void> {
    const client = await this.pool.connect();
    try {
      await client.query('BEGIN');
      try {
        await fn(client);
        await client.query('COMMIT');
      } catch (e) {
        console.error('Error during transaction:', e);
        await client.query('ROLLBACK');
        throw e; // Re-throw the error to propagate it
      }
    } finally {
      client.release();
    }
  }

  /**
   * Shuts down the connection pool.
   */
  async shutdown(): Promise<void> {
    await this.pool.end();
  }
}