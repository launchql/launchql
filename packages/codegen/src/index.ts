import { generateCode } from './codegen';
import getIntrospectionRows, { GetIntrospectionRowsOptions } from './introspect';
import { DatabaseObject } from './types';

(async () => {
  const options: GetIntrospectionRowsOptions = {
    connectionString: 'postgresql://postgres:password@localhost:5432/mydb',
    introspectionOptions: {
      pgLegacyFunctionsOnly: false,
      pgIgnoreRBAC: true,
    },
    namespacesToIntrospect: ['collections_public'],
    includeExtensions: false,
  };

  try {
    // Fetch introspection rows
    const rows: DatabaseObject[] = await getIntrospectionRows(options);
    console.log('Introspection Rows Fetched:', rows);

    // Generate TypeScript code
    const codegenOptions = {
      includeTimestamps: true,
      includeUUID: true,
    };
    const generatedCode = generateCode(rows, codegenOptions);

    console.log('Generated TypeScript Code:');
    console.log(generatedCode);
  } catch (error) {
    console.error('Failed to fetch introspection rows or generate code:', error);
  }
})();
