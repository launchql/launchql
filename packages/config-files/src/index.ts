export * from './types';
export * from './plan';
export * from './config';
export * from './sql';
export * from './extension';
export * from './sql-scripts';