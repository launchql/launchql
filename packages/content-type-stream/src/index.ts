export * from './content-stream';
export * from './content-type-stream';