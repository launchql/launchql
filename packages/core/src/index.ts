export * from './core/class/launchql';
export * from './export/export-meta';
export * from './export/export-migrations';
export * from './extensions/extensions';
export * from './modules/modules';
export * from './packaging/package';
export * from './packaging/transform';
export * from './resolution/deps';
export * from './resolution/resolve';
export * from './workspace/paths';
export * from './workspace/utils';

// Export project-files functionality (now integrated into core)
export * from './files';
export { cleanSql } from './migrate/clean';
export { LaunchQLMigrate } from './migrate/client';
export { 
  DeployOptions, 
  DeployResult, 
  MigrateChange, 
  MigratePlanFile, 
  RevertOptions, 
  RevertResult, 
  StatusResult,
  VerifyOptions, 
  VerifyResult} from './migrate/types';
export { hashFile, hashString } from './migrate/utils/hash';
export { executeQuery,TransactionContext, TransactionOptions, withTransaction } from './migrate/utils/transaction';
