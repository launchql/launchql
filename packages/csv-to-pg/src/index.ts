// @ts-nocheck
export * from './parse';
export * from './parser';
