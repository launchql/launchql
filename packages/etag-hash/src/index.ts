var createHash = require('crypto').createHash;

class ETagHash {
  constructor(partSizeInMb = 5) {
    this.partSizeInBytes = partSizeInMb * 1024 * 1024;
    this.sums = [createHash('md5')];
    this.part = 0;
    this.bytes = 0;
  }

  update(chunk) {
    const len = chunk.length;

    if (this.bytes + len < this.partSizeInBytes) {
      this.sums[this.part].update(chunk);
      this.bytes += len;
    } else {
      const bytesNeeded = this.partSizeInBytes - this.bytes;
      this.sums[this.part].update(chunk.slice(0, bytesNeeded));
      this.part++;
      this.sums.push(createHash('md5'));
      this.bytes = len - bytesNeeded;
      this.sums[this.part].update(chunk.slice(bytesNeeded, len));
    }

    return this;
  }

  digest() {
    if (!this.part) {
      return this.sums[0].digest('hex');
    }

    const checksum = this.sums.map(s => s.digest('hex')).join('');
    const final = createHash('md5')
      .update(Buffer.from(checksum, 'hex'))
      .digest('hex');

    return `${final}-${this.part + 1}`;
  }
}

module.exports = ETagHash;
module.exports.createHash = (...args) => {
  return new ETagHash(...args);
};
