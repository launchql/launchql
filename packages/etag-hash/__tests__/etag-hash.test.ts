import { createHash } from '../src';

describe('ETag', () => {
  it('ETag hash', async () => {
    const res: Record<string, string> = {};
    const strings = ['Hello World', 'Another String'];

    for (let i = 0; i < strings.length; i++) {
      const str = strings[i];
      res[str] = createHash()
        .update(Buffer.from(str, 'utf8'))
        .digest();
    }

    expect(res).toMatchSnapshot();
  });

  it('Large ETag hash', async () => {
    const hash = createHash();

    const SZ = 5 * 1024 * 1024;
    for (let i = 0; i < SZ; i++) {
      hash.update(Buffer.from(String(i), 'utf8'));
    }

    const res = hash.digest();
    expect(res).toMatchSnapshot();
  });
});
