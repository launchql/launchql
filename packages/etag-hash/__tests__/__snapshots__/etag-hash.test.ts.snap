// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ETag ETag hash 1`] = `
{
  "Another String": "228dd4caf295a9d5b0a4e54b6be9f6f8",
  "Hello World": "b10a8db164e0754105b7a99be72e3fe5",
}
`;

exports[`ETag Large ETag hash 1`] = `"b5f547c711c4cf2d50088473dcd4fa92-7"`;
